### PR TITLE
Refactor CodeLocation and type its properties.

### PR DESCRIPTION
**Pull request by ltfish**

### CODE REVIEW

**Review by Pamplemousse — COMMENTED** [review]
Looks much better!

I have a question / remark though: 
Should context accept `None` as a value at all? Shouldn't its "default" be `()` (`None` as parameter of `__init__`, but set to `()` during instanciation)?

File: angr/code_location.py [hunk 1]
```diff
@@ -7,23 +7,27 @@ class CodeLocation:
...
         """
 
-        self.block_addr = block_addr
-        self.stmt_idx = stmt_idx
+        self.block_addr: int = block_addr
+        self.stmt_idx: int = stmt_idx
         self.sim_procedure = sim_procedure
-        self.ins_addr = ins_addr
+        self.ins_addr: Optional[int] = ins_addr
+        # sanitization: if context is an empty tuple, we store a None instead
```
[the remark this answers is not in the document]
Comment (ltfish) [reply angr/code_location.py:29]: One less object.

**Comment by ltfish** [comment]
> 
> 
> Looks much better!
> 
> I have a question / remark though:
> Should context accept `None` as a value at all? Shouldn't its "default" be `()` (`None` as parameter of `__init__`, but set to `()` during instanciation)?

I feel we should allow context-less `CodeLocation` to exist.

**Comment by Pamplemousse** [comment]
> I feel we should allow context-less CodeLocation to exist.

Yeah, I tend to agree.
However, because `()` is stored as `None`, there is no distinction possible between "empty context" and "context-less"...

**Comment by ltfish** [comment]
> > I feel we should allow context-less CodeLocation to exist.
> 
> Yeah, I tend to agree.
> However, because `()` is stored as `None`, there is no distinction possible between "empty context" and "context-less"...

So the problem seems to be that we are passing in `None` at places where we should actually be passing in empty tuples. I’ll look into this issue.

File: angr/code_location.py (hunk 1 above)
Comment (Pamplemousse) [inline angr/code_location.py:29]: This comment is not needed anymore!

**Review by Pamplemousse — CHANGES_REQUESTED** [review]

File: angr/code_location.py (hunk 1 above)
Comment (ltfish) [reply angr/code_location.py]: Fixed!

**Review by Pamplemousse — APPROVED** [review]

**Pull request merged**